### PR TITLE
Add additional checking around memcpy usage

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @bartonjs, @morganbr, @janvorli

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test this please (Jenkins didn't respond)

File: src/Native/Unix/Common/pal_safecrt.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        if (dst == nullptr)
+        {
+            return EINVAL;
+        }
+
+        if (src == nullptr || sizeInBytes < count)
+        {
+            memset(dst, 0, sizeInBytes);
+            return EINVAL;
```
Comment (bartonjs) [inline src/Native/Unix/Common/pal_safecrt.h]: EINVAL for `src == nullptr`, ERANGE for `sizeInBytes < count`.  (https://msdn.microsoft.com/en-us/library/wes2t00f.aspx#Anchor_2)

Comment (stephentoub) [inline src/Native/Unix/Common/pal_safecrt.h]: I can change it, but I was following what coreclr has, which is using EINVAL here:
https://github.com/dotnet/coreclr/blob/32f0f9721afb584b4a14d69135bea7ddc129f755/src/pal/src/safecrt/memcpy_s.cpp#L75

Comment (bartonjs) [inline src/Native/Unix/Common/pal_safecrt.h]: I'm pretty sure that the line above that is a macro which returns the second argument (particularly given the comment)

`_VALIDATE_RETURN_ERRCODE(sizeInBytes >= count, ERANGE);`

Comment (stephentoub) [inline src/Native/Unix/Common/pal_safecrt.h]: Ah, of course, you're right.

**Comment by mellinoe** [comment]
NetworkInformation changes LGTM

**Comment by stephentoub** [comment]
@dotnet-bot test Innerloop Windows_NT Debug Build and Test please (https://github.com/dotnet/corefx/issues/13025)

**Comment by bartonjs** [comment]
The CentOS debug leg coredumped.

**Comment by stephentoub** [comment]
> The CentOS debug leg coredumped.

Yeah, one of the asserts caught an issue in networking; I'm fixing it.

**Comment by stephentoub** [comment]
@bartonjs, @ericeil, I added another commit that should address the assert that was firing.  Could you take a look?

File: src/Native/Unix/System.Native/pal_networking.cpp [hunk 2]
```diff
@@ -1647,13 +1647,14 @@ extern "C" Error SystemNative_ReceiveMessage(intptr_t socket, MessageHeader* mes
...
     while (CheckInterrupted(res = recvmsg(fd, &header, socketFlags)));
 
+    assert(header.msg_name == messageHeader->SocketAddress); // should still be the same location as set in ConvertMessageHeaderToMsghdr
+    assert(header.msg_control == messageHeader->ControlBuffer);
+
     assert(static_cast<int32_t>(header.msg_namelen) <= messageHeader->SocketAddressLen);
     messageHeader->SocketAddressLen = Min(static_cast<int32_t>(header.msg_namelen), messageHeader->SocketAddressLen);
-    memcpy_s(messageHeader->SocketAddress, static_cast<size_t>(messageHeader->SocketAddressLen), header.msg_name, static_cast<size_t>(messageHeader->SocketAddressLen));
-
+
```
Comment (bartonjs) [inline src/Native/Unix/System.Native/pal_networking.cpp]: Nit: Added positive whitespace on this line

**Comment by bartonjs** [comment]
Probably LGTM. I slightly wonder if the asserts should be full runtime checks that fail... but it's probably fine. :smile:

**Comment by stephentoub** [comment]
Thanks, Jeremy.

**Pull request merged**